universal-login-relayer/lib/services/messages/PendingMessages.ts

```diff
@@ -1,62 +1,68 @@
 
   constructor(private wallet : Wallet, private messagesStore: IPendingMessagesStore) {
   }
 
-  isPresent(messageHash : string) {
+  async isPresent(messageHash : string) {
     return this.messagesStore.isPresent(messageHash);
   }
 
   async add(message: SignedMessage) : Promise<string> {
-    const hash = calculateMessageHash(message);
-    if (!this.isPresent(hash)) {
-      this.messagesStore.add(hash, new PendingMessage(message.from, this.wallet));
+    const messageHash = calculateMessageHash(message);
+    if (!await this.isPresent(messageHash)) {
+      const pendingMessage = newPendingMessage(message.from);
+      await this.messagesStore.add(messageHash, pendingMessage);
     }
-    await this.addSignatureToPendingMessage(hash, message);
-    return hash;
+    await this.addSignatureToPendingMessage(messageHash, message);
+    return messageHash;
   }
 
-  private async addSignatureToPendingMessage(hash: string, message: SignedMessage) {
-    const pendingMessage = this.messagesStore.get(hash);
+  private async addSignatureToPendingMessage(messageHash: string, message: SignedMessage) {
+    const pendingMessage = await this.messagesStore.get(messageHash);
     this.ensureCorrectTransactionHash(pendingMessage.transactionHash);
-    if (pendingMessage.containSignature(message.signature)) {
+    if (await this.messagesStore.containSignature(messageHash, message.signature)) {
       throw new DuplicatedSignature();
     }
     const key = getKeyFromHashAndSignature(
       calculateMessageHash(message),
       message.signature
     );
-    const keyPurpose = await pendingMessage.walletContract.getKeyPurpose(key);
+    const walletContract = new Contract(pendingMessage.walletAddress, WalletContract.interface, this.wallet);
+    const keyPurpose = await walletContract.getKeyPurpose(key);
     if (keyPurpose.eq(INVALID_KEY)) {
       throw new InvalidSignature('Invalid key purpose');
     }
-    this.messagesStore.addSignature(hash, message.signature);
+    await this.messagesStore.addSignature(messageHash, message.signature);
   }
 
-  async getStatus(hash: string) {
-    return this.messagesStore.getStatus(hash);
+  async getStatus(messageHash: string) {
+    return this.messagesStore.getStatus(messageHash, this.wallet);
   }
 
-  getMessageWithSignatures(message: SignedMessage, hash: string) : SignedMessage {
-    return  { ...message, signature: this.messagesStore.get(hash).getConcatenatedSignatures()};
+  async getMessageWithSignatures(message: SignedMessage, messageHash: string) : Promise<SignedMessage> {
+    const collectedSignatureKeyPairs = await this.messagesStore.getCollectedSignatureKeyPairs(messageHash);
+    const sortedSignatureKeyPairs = sortSignatureKeyPairsByKey([...collectedSignatureKeyPairs]);
+    const sortedSignatures = sortedSignatureKeyPairs.map((value: CollectedSignatureKeyPair) => value.signature);
+    const signature = concatenateSignatures(sortedSignatures);
+    return  { ...message, signature};
```

Comment:
<3
Finally, this is exactly how logic should be split between abstraction layers and util function